### PR TITLE
add Env to NME Preview

### PR DESCRIPTION
**Pull request by onekit-boss**
We can custom Env, Image and toggle Env / Background color now.

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/14398/merge

Test environment: 
https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14398/merge/index.html

To test a playground add it to the URL, for example:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14398/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/14398/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/14398/merge
https://gui.babylonjs.com/?snapshot=refs/pull/14398/merge
https://nme.babylonjs.com/?snapshot=refs/pull/14398/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/14398/merge#BCU1XR#0

**Review by Popov72 — COMMENTED** [review]
Just to be sure we are ok with the GUI, here's how it appears:
image

cc @PatrickRyanMS

File: packages/tools/nodeEditor/src/components/preview/previewAreaComponent.tsx [hunk 1]
```diff
@@ -155,6 +156,18 @@ export class PreviewAreaComponent extends React.Component<IPreviewAreaComponentP
                             >
                                 <img src={directionalLeft} alt="" />
                             </div>
+                            <div
+                                title="Turn on/off background"
```
[the remark this answers is not in the document]
Comment (onekit-boss) [reply packages/tools/nodeEditor/src/components/preview/previewAreaComponent.tsx]: ok. I will rename it

**Comment by sebavan** [comment]
LGTM if @PatrickRyanMS is all good with it :-)

**Review by deltakosh — APPROVED** [review]

**Comment by sebavan** [comment]
Also one last thought for @RaananW to be sure importing .env would not break any kindz of buildzzzz 🐝🐝🐝🐝🐝

**Comment by RaananW** [comment]
> Also one last thought for @RaananW to be sure importing .env would not break any kindz of buildzzzz 🐝🐝🐝🐝🐝

Works fine, as this is just a static file in the public directory.
Might cause an issue when used as a package, as you will need to deliver the file yourself.

**Review by sebavan — CHANGES_REQUESTED** [review]
Nice thanks a lot !!! a last change and I can merge it in :-) Could you make "https://assets.babylonjs.com/environments/environmentSpecular.env" a public static so that users can point it at their own URL if need be :-)

**Comment by onekit-boss** [comment]
> > Nice thanks a lot !!! a last change and I can merge it in :-) Could you make "https://assets.babylonjs.com/environments/environmentSpecular.env" a public static so that users can point it at their own URL if need be :-)
> 
> yes. Now it is a exists URL of babylon assets (https://assets.babylonjs.com), not a file .

I use this URL as other source codes of babylon.js .your can find the URL in bellow files:
packages\dev\core\src\Helpers\environmentHelper.ts
packages\tools\sandbox\src\tools\environmentTools.ts

**Comment by sebavan** [comment]
yes but in environmentHelper users can override it here https://github.com/BabylonJS/Babylon.js/blob/master/packages/dev/core/src/Helpers/environmentHelper.ts#L344 and sandbox is not meant to be used as a framework. 

Users needs to be able to control where they download assets from.

**Comment by onekit-boss** [comment]
> yes but in environmentHelper users can override it here https://github.com/BabylonJS/Babylon.js/blob/master/packages/dev/core/src/Helpers/environmentHelper.ts#L344 and sandbox is not meant to be used as a framework.
> 
> Users needs to be able to control where they download assets from.

your mean is. ........"user need custom a env of the NME Preview window?" . I has add a "upload button" on the right of "model ComboBox" in NME Preview window. so user can override the default env "https://assets.babylonjs.com/environments/environmentSpecular.env".

**Comment by sebavan** [comment]
Oh no, sorry, I just mean if you move it in a static public field or a let in the module so a developer consuming the npm package can change it before loading nme.

**Comment by onekit-boss** [comment]
> a public static

or...... your mean is a public static  filed  of the class?

**Comment by sebavan** [comment]
something like this:

```
public static DefaultEnvironmentURL = "https://assets.babylonjs.com/environments/environmentSpecular.env"';

private _refreshPreviewMesh(force?: boolean) {
        switch (this._globalState.envType) {
            case PreviewType.Room:
                this._hdrTexture = new CubeTexture(DefaultEnvironmentURL , this._scene);
                if (this._hdrTexture) {
                    this._prepareBackgroundHDR();
                }
                break;
            case PreviewType.Custom: {
                const blob = new Blob([this._globalState.envFile], { type: "octet/stream" });
                const reader = new FileReader();
                reader.onload = (evt) => {
                    const dataurl = evt.target!.result as string;
                    this._hdrTexture = new CubeTexture(dataurl, this._scene, undefined, false, undefined, undefined, undefined, undefined, undefined, ".env");
                    this._prepareBackgroundHDR();
                };
                reader.readAsDataURL(blob);
                break;
            }
        }
```

**Comment by onekit-boss** [comment]
> something like this:
> 
> ```
> public static DefaultEnvironmentURL = "https://assets.babylonjs.com/environments/environmentSpecular.env"';
> 
> private _refreshPreviewMesh(force?: boolean) {
>         switch (this._globalState.envType) {
>             case PreviewType.Room:
>                 this._hdrTexture = new CubeTexture(DefaultEnvironmentURL , this._scene);
>                 if (this._hdrTexture) {
>                     this._prepareBackgroundHDR();
>                 }
>                 break;
>             case PreviewType.Custom: {
>                 const blob = new Blob([this._globalState.envFile], { type: "octet/stream" });
>                 const reader = new FileReader();
>                 reader.onload = (evt) => {
>                     const dataurl = evt.target!.result as string;
>                     this._hdrTexture = new CubeTexture(dataurl, this._scene, undefined, false, undefined, undefined, undefined, undefined, undefined, ".env");
>                     this._prepareBackgroundHDR();
>                 };
>                 reader.readAsDataURL(blob);
>                 break;
>             }
>         }
> ```

thanks. I got it :)

**Review by sebavan — APPROVED** [review]

**Comment by PatrickRyanMS** [comment]
> Just to be sure we are ok with the GUI, here's how it appears: image
> 
> cc @PatrickRyanMS

sorry for the late response, I have been out of the office. The only thing I see is that the icon for environment is a little small in comparison to the rest on the bar. It should be sized to fit with the size of the other icons. If needed, I can provide a new icon, but it would be preferable to use the same svg if possible. Something like this:

image

**Comment by sebavan** [comment]
@PatrickRyanMS it looks real strange zoomed in like it would be a bold font :-( could you create another svg ? like the other ones in the bar ?

**Comment by PatrickRyanMS** [comment]
I can make a new svg to fit with the current icons. Give me a few minutes.

**Comment by PatrickRyanMS** [comment]
@sebavan, here's the updated icon. This is saved at the same size as the rest of the icons, so the css should be the same:
[environment_24_regular.zip](https://github.com/BabylonJS/Babylon.js/files/12860764/environment_24_regular.zip)

image

**Pull request merged**